### PR TITLE
FileParameter as new ParameterConfig to handle files in requests.

### PR DESCRIPTION
**Pull request by tsoenen**

### CODE REVIEW

File: test/test_jsonplaceholder.py
Comment (swinkels-at-basf) [inline test/test_jsonplaceholder.py:191]: Extreme nitpick: the test doesn't close the open file. This should only be a theoretical problem because the file will be closed when the last reference to its object is deleted.

**Review by swinkels-at-basf — APPROVED** [review]
Some minor comments that are definately not enough to hold off a merge. All unit tests run and I assume ;) you have tested it yourself, so LGTM.

File: qrest/resource.py
Comment (tsoenen) [reply qrest/resource.py:475]: I think the purpose of these `extra_` parameters is to allow for predefined values for parameters. E.g., a scenario where an endpoint has three query parameters, but you only want to 'offer' one to the user and set two in the ResourceConfig.

Maybe a more elegant approach would be to extend the ConfigParameter with a `predefined` attribute. Such a parameter would not be availble to the user to be set, but would go through the same validation as the other parameters, making the `extra_` parameters and their checks obsolete.

File: test/test_configuration.py
Comment (tsoenen) [reply test/test_configuration.py:80]: Will keep it in mind for future PRs

File: test/test_jsonplaceholder.py
Comment (tsoenen) [reply test/test_jsonplaceholder.py:191]: Agreed, missed this one.

File: qrest/resource.py
Comment (swinkels-at-basf) [reply qrest/resource.py:475]: > I think the purpose of these extra_ parameters is to allow for predefined values for parameters.

There might be an idea behind it, or maybe an initial idea that in the end wasn't full realized. @JorisBenschop do you know why the these parameters are there?

Comment (JorisBenschop) [reply qrest/resource.py:475]: I think this was added to deal with an unpleasant design aspect of Phoenix, where a POST would have the pagination parameters in the GET. Now the pagination parameters are not meant to be provided to the user (and should also not show in the parameter lists or other introspection aspects), as the internal pagination code should deal with this. So these pagination parameters are injected into the parameter dict after the user has pressed "enter".

Comment (JorisBenschop) [reply qrest/resource.py:475]: not sure if that means we even need an extra_file though

Comment (swinkels-at-basf) [reply qrest/resource.py:475]: Now I understand, `extra_request` isn't used in qrest itself but is used from (what used to be) phoenix-rest to support a paged resource / response. Thanks for the clarification.

Comment (JorisBenschop) [reply qrest/resource.py:475]: At least we can conclude that i did not document this sufficiently

**Pull request merged**
